src/dir.rs

```diff
@@ -0,0 +1,167 @@
+use {Error, Result};
+use errno::Errno;
+use libc;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, fmt, ptr};
+
+#[cfg(target_os = "linux")]
+use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+/// An open directory.
+///! This is a lower-level interface than `std::fs::ReadDir`. Notable differences:
+///!    * can be opened from a file descriptor (as returned by `openat`, perhaps before knowing
+///!      if the path represents a file or directory).
+///!    * implements `AsRawFd`, so it can be passed to `fstat`, `openat`, etc.
+///!    * can be iterated through multiple times without closing and reopening the file
+///!      descriptor. Each iteration rewinds when finished.
+///!    * returns entries for `.` (current directory) and `..` (parent directory).
+///!    * returns entries' names as a `CStr` (no allocation or conversion beyond whatever libc
+///!      does).
+pub struct Dir(ptr::NonNull<libc::DIR>);
+
+impl Dir {
+    #[inline]
+    pub fn from<F: IntoRawFd>(fd: F) -> Result<Self> {
+        Dir::from_fd(fd.into_raw_fd())
+    }
+
+    /// Converts from a file descriptor, closing it on success or failure.
+    pub fn from_fd(fd: RawFd) -> Result<Self> {
+        match ptr::NonNull::new(unsafe { libc::fdopendir(fd) }) {
+            None => {
+                let e = Error::last();
+                unsafe { libc::close(fd) };
```

Comment:
`Dir` takes ownership of the file descriptor, and closes it on `Drop`.  Unfortunately, `RawFd` is defined as an alias, so there's no way to enforce that the caller doesn't hold onto a copy.  But could you at least warn the caller in the documentation?  Bad stuff will happen if he accesses fd outside of the `Dir` interface.

Reply:
`AsRawFd::as_raw_fd` has such a warning:

> This method does not pass ownership of the raw file descriptor to the caller. The descriptor is only guaranteed to be valid while the original object has not yet been destroyed.

but I'll add one to `Dir`'s docs as well where I mention the `AsRawFd` implementation.